File: Olympiad/COCI/official/2008/regional_solutions/kuhar.cpp
#include <cstdio>

const int inf = 1000000000;

int n, k;
int x[100], y[100];
int cm[100], sm[100];
int cv[100], sv[100];

int nabavi( int grama, int cm, int sm, int cv, int sv ) {
   int rez = inf;
   int velikih = 0;
   do {
      int malih = (grama - velikih * sv + sm-1) / sm >? 0;
      rez <?= malih * cm + velikih * cv;
   } while( velikih++ * sv <= grama );
   return rez;
}

bool moguce( int porcija ) {
   int cijena = 0;
   for( int i = 0; i < n; ++i )
      cijena += nabavi( x[i] * porcija - y[i], cm[i], sm[i], cv[i], sv[i] );
   return cijena <= k;
}

int main( void ) {
   scanf( "%d%d", &n, &k );
   for( int i = 0; i < n; ++i )
      scanf( "%d%d%d%d%d%d", &x[i], &y[i], &sm[i], &cm[i], &sv[i], &cv[i] );

   int lo = 0, hi = 1;
   while( moguce( hi ) ) hi = hi * 2;

   while( lo < hi ) {
      int mid = (lo+hi+1) / 2;
      if( moguce( mid ) ) lo = mid; else hi = mid-1;
   }

   printf( "%d\n", lo );

   return 0;
}

File: Olympiad/COCI/official/2008/regional_solutions/kuhar.pas
program kuhar;

const inf = 1000000000;

var
   i, n, k : longint;
   x, y : array[1..100] of longint;
   cm, sm : array[1..100] of longint;
   cv, sv : array[1..100] of longint;
   lo, hi, mid : longint;

function nabavi( grama, cm, sm, cv, sv : longint ) : longint;
var
   rez, malih, velikih : longint;
begin
   rez := inf;

   velikih := 0;
   repeat
      malih := (grama - velikih * sv + sm-1) div sm;
      if malih < 0 then malih := 0;

      if malih * cm + velikih * cv < rez then
         rez := malih * cm + velikih * cv;

      velikih := velikih + 1;
   until (velikih-1) * sv > grama;

   nabavi := rez;
end;

function moguce( porcija : longint ) : boolean;
var
   cijena, i : longint;
begin
   cijena := 0;
   for i := 1 to n do
      cijena := cijena + nabavi( x[i] * porcija - y[i], cm[i], sm[i], cv[i], sv[i] );

   if cijena <= k then moguce := true else moguce := false;
end;

begin
   readln( n, k );
   for i := 1 to n do
      readln( x[i], y[i], sm[i], cm[i], sv[i], cv[i] );

   lo := 0;
   hi := 1;
   while moguce( hi ) do hi := hi * 2;

   while lo < hi do begin
      mid := (lo+hi+1) div 2;
      if moguce( mid ) then lo := mid else hi := mid-1;
   end;

   writeln( lo );
end.
